File: queens/solver/hint.py
"""Top-level hint search, candidate bookkeeping, and the backtracking solver —
mirrors ``sudoku/solver/hint.py``'s shape.

``find_hint`` returns the single simplest applicable step, walking
``TECHNIQUES`` (see ``techniques.py``) in order. ``working_candidates`` builds
the candidate grid it reasons over fresh from the board each time it's called:
Queens has no pencil-mark-vs-derived-candidate reconciliation to do the way
Sudoku does, since a Mark already *is* the board's record of "ruled out" —
there's no separate mark set to intersect against legal candidates.

``solve`` (backtracking) was added in issue #4; this pass (#6) adds the
elimination-propagation bookkeeping and hint search alongside it.
"""

from __future__ import annotations

from typing import Optional

from ..model import EMPTY, MARKED, QUEEN, Board
from .techniques import CandGrid, Coord, Hint, TECHNIQUES


def working_candidates(board: Board) -> CandGrid:
    """Build the candidate grid the engine reasons over: every empty cell not
    ruled out by an existing queen's row, column, region (``Board.peers``) or
    8-neighbors (``Board.neighbors``) — Queens' elimination-propagation
    bookkeeping. Marked cells are excluded automatically since they aren't
    ``EMPTY``; a Mark carries no extra information beyond "not a candidate"
    (see ``queens/CONTEXT.md``).
    """
    blocked: set[Coord] = set()
    for r, c in board.queen_cells():
        blocked.update(board.peers(r, c))
        blocked.update(board.neighbors(r, c))
    return {
        (r, c)
        for r, c in board.coords()
        if board.state(r, c) == EMPTY and (r, c) not in blocked
    }


def find_hint(board: Board, cg: Optional[CandGrid] = None) -> Optional[Hint]:
    """First applicable technique, simplest first. ``None`` if invalid,
    solved, or stuck (no implemented technique applies)."""
    if not board.is_valid():
        return None
    if cg is None:
        cg = working_candidates(board)
    for technique in TECHNIQUES:
        hint = technique(board, cg)
        if hint is not None:
            return hint
    return None


def apply_to_candidates(board: Board, cg: CandGrid, hint: Hint) -> CandGrid:
    """Apply a hint to a candidate grid in place and return it.

    Takes ``board`` in addition to ``(cg, hint)`` — unlike Sudoku's
    ``apply_to_candidates(cg, hint)`` — because a cell's peers/neighbors
    depend on board size and region layout here, which are instance data
    (``Board.n``, per-cell ``region``) rather than a fixed 9x9 layout ``sudoku``
    can compute from coordinates alone.

    A placement drops the placed cell from ``cg`` and propagates elimination
    to its peers and neighbors, exactly like ``working_candidates`` does for
    a queen already on the board. An elimination (issues #7/#8) just drops
    the named cells.
    """
    if hint.action == "place":
        (r, c) = hint.cells[0]
        cg.discard((r, c))
        cg -= board.peers(r, c)
        cg -= board.neighbors(r, c)
    else:
        for cell in hint.cells:
            cg.discard(cell)
    return cg


def apply_hint(board: Board, hint: Hint) -> Board:
    """Return a copy of ``board`` with ``hint`` applied (a queen for a
    placement, a mark for an elimination)."""
    new = Board.from_dict(board.to_dict())
    if hint.action == "place":
        (r, c) = hint.cells[0]
        new.set_state(r, c, QUEEN)
    else:
        for (r, c) in hint.cells:
            new.set_state(r, c, MARKED)
    return new


def solve_with_techniques(board: Board) -> tuple[Board, list[Hint], bool]:
    """Repeatedly apply ``find_hint`` until solved or stuck.

    Returns ``(final_board, steps, solved)``. Operates on a persistent
    candidate grid so eliminations accumulate, and writes placements back to
    the board as it goes (mirroring ``sudoku.solver.hint.solve_with_techniques``).
    """
    work = Board.from_dict(board.to_dict())
    cg = working_candidates(work)
    steps: list[Hint] = []
    while True:
        hint = find_hint(work, cg)
        if hint is None:
            break
        steps.append(hint)
        if hint.action == "place":
            (r, c) = hint.cells[0]
            work.set_state(r, c, QUEEN)
        else:
            for (r, c) in hint.cells:
                work.set_state(r, c, MARKED)
        apply_to_candidates(work, cg, hint)
    return work, steps, work.is_solved()


def solve(board: Board) -> Optional[Board]:
    """Backtracking solver. Returns a solved copy, or ``None`` if unsolvable."""
    work = Board.from_dict(board.to_dict())
    if not work.is_valid():
        return None
    if _backtrack(work, 0):
        return work
    return None


def _backtrack(board: Board, row: int) -> bool:
    """Recurse row-by-row: exactly one queen must land in each row, so each
    level of recursion picks that row's column (or, if the row already has a
    queen from the starting board, moves straight on)."""
    if row == board.n:
        return True
    if any(board.state(row, c) == QUEEN for c in range(board.n)):
        return _backtrack(board, row + 1)
    for col in range(board.n):
        if board.state(row, col) != EMPTY:
            continue  # marked cells can never receive a queen
        if not _can_place(board, row, col):
            continue
        board.set_state(row, col, QUEEN)
        if _backtrack(board, row + 1):
            return True
        board.set_state(row, col, EMPTY)
    return False


def _can_place(board: Board, r: int, c: int) -> bool:
    region = board.region(r, c)
    for pr, pc in board.queen_cells():
        if pc == c:
            return False
        if region is not None and board.region(pr, pc) == region:
            return False
        if abs(pr - r) <= 1 and abs(pc - c) <= 1:
            return False
    return True

File: web/queens/hint.ts
// Top-level hint search and Candidate bookkeeping. (The backtracking solver
// lives in solver.ts.)
//
// `findHint` returns the single simplest applicable step, walking `TECHNIQUES`
// (see techniques.ts) in order. `workingCandidates` builds the Candidate grid
// it reasons over fresh from the board each time it is called: Queens has no
// pencil-mark-versus-derived-candidate reconciliation to do the way Sudoku
// does, since a Mark already *is* the board's record of "ruled out" — there is
// no separate mark set to intersect against legal candidates.

import { Board, EMPTY, MARKED, QUEEN, cellName, type Coord } from "./model.ts";
import { TECHNIQUES, type CandGrid, type Hint } from "./techniques.ts";

/** A hint in the shape the page uses, where a Cell is `{r, c}` rather than a
 * pair. The catalogue itself never sees this — it is the page's shape, and it
 * lives here with the rest of what the page calls. */
export type WireHint = Omit<Hint, "cells"> & { cells: { r: number; c: number }[] };

export function hintToWire(hint: Hint): WireHint {
  return { ...hint, cells: hint.cells.map(([r, c]) => ({ r, c })) };
}

/**
 * The Candidate grid the engine reasons over: every empty cell not ruled out by
 * an existing queen's row, column, region (`Board.peers`) or 8-neighbors
 * (`Board.neighbors`) — Queens' elimination-propagation bookkeeping. Marked
 * cells are excluded automatically since they are not empty; a Mark carries no
 * extra information beyond "not a candidate" (see web/queens/CONTEXT.md).
 */
export function workingCandidates(board: Board): CandGrid {
  const blocked = new Set<number>();
  for (const [r, c] of board.queenCells()) {
    for (const i of board.peers(r, c)) blocked.add(i);
    for (const i of board.neighbors(r, c)) blocked.add(i);
  }
  const cg: CandGrid = new Set();
  for (const [r, c] of board.coords()) {
    const i = board.idx(r, c);
    if (board.state(r, c) === EMPTY && !blocked.has(i)) cg.add(i);
  }
  return cg;
}

/** First applicable technique, simplest first. `null` if invalid, solved, or
 * stuck (no implemented technique applies). */
export function findHint(board: Board, cg?: CandGrid): Hint | null {
  if (!board.isValid()) return null;
  const grid = cg ?? workingCandidates(board);
  for (const technique of TECHNIQUES) {
    const hint = technique(board, grid);
    if (hint !== null) return hint;
  }
  return null;
}

/**
 * Apply a hint to a Candidate grid in place and return it.
 *
 * Takes `board` in addition to `(cg, hint)` because a cell's peers/neighbors
 * depend on board size and region layout, which are instance data (`Board.n`,
 * per-cell `region`). Sudoku's counterpart takes a board for the same reason —
 * its peers became board-dependent once Killer cages entered the model.
 *
 * A placement drops the placed cell from `cg` and propagates elimination to its
 * peers and neighbors, exactly like `workingCandidates` does for a queen
 * already on the board. An elimination (issues #7/#8) just drops the named
 * cells.
 */
export function applyToCandidates(board: Board, cg: CandGrid, hint: Hint): CandGrid {
  if (hint.action === "place") {
    const [r, c] = hint.cells[0];
    cg.delete(board.idx(r, c));
    for (const i of board.peers(r, c)) cg.delete(i);
    for (const i of board.neighbors(r, c)) cg.delete(i);
  } else {
    for (const [r, c] of hint.cells) cg.delete(board.idx(r, c));
  }
  return cg;
}

/** A copy of `board` with `hint` applied (a queen for a placement, a mark for
 * an elimination). */
export function applyHint(board: Board, hint: Hint): Board {
  const next = Board.fromWire(board.toWire());
  if (hint.action === "place") {
    const [r, c] = hint.cells[0];
    next.setState(r, c, QUEEN);
  } else {
    for (const [r, c] of hint.cells) next.setState(r, c, MARKED);
  }
  return next;
}

export interface TechniqueRun {
  board: Board;
  steps: Hint[];
  solved: boolean;
}

/**
 * Repeatedly apply `findHint` until solved or stuck.
 *
 * Operates on a persistent Candidate grid so eliminations accumulate, and
 * writes placements back to the board as it goes.
 */
export function solveWithTechniques(board: Board, cg?: CandGrid): TechniqueRun {
  const work = Board.fromWire(board.toWire());
  const grid = cg ?? workingCandidates(work);
  const steps: Hint[] = [];
  for (;;) {
    const hint = findHint(work, grid);
    if (hint === null) break;
    steps.push(hint);
    if (hint.action === "place") {
      const [r, c] = hint.cells[0] as Coord;
      work.setState(r, c, QUEEN);
    } else {
      for (const [r, c] of hint.cells) work.setState(r, c, MARKED);
    }
    applyToCandidates(work, grid, hint);
  }
  return { board: work, steps, solved: work.isSolved() };
}

/** The gentlest reveal: which Unit or Cell to look at, without saying what to
 * do.
 *
 * Queens keeps its own implementation of this rather than reusing Sudoku's, so
 * the two engines stay independent per ADR 0001. */
export function nudge(hint: Hint): string {
  if (hint.units.length > 0) return `Look at ${hint.units[0]}.`;
  const [r, c] = hint.cells[0];
  return `Look at ${cellName(r, c)}.`;
}
